src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -185,121 +166,188 @@ def __enter__(self):
+                    message_format=f"ProcessPool failed with exception: {e}",
+                    target=ErrorTarget.EXECUTOR,
+                ) from e
 
-    def _is_process_alive(self, process_id):
-        return psutil.pid_exists(process_id)
+        if self._batch_timeout_expired(self._batch_start_time):
+            # Send terminate signal to all threads and wait for them to exit.
+            self._terminate_tasks()
+            # Wait for up to 10s for thread termination, aiming to ensure that
+            # the line results in the result dict are as complete as possible.
+            start_time = datetime.utcnow()
+            while not self._timeout_expired(start_time, self._THREAD_TERMINATED_TIMEOUT):
+                if self._all_tasks_ready():
+                    break
+            # Set the timeout flag to True and log the warning.
+            self._is_timeout = True
+            bulk_logger.warning(f"The batch run timed out, with {len(self._result_dict)} line results processed.")
+        return [self._result_dict[key] for key in sorted(self._result_dict)]
 
-    def _handle_output_queue_messages(self, output_queue: Queue, result_list: List[LineResult]):
-        try:
-            message = output_queue.get(timeout=1)
-            if isinstance(message, LineResult):
-                message = self._process_multimedia(message)
-                result_list.append(message)
-                return message
-            elif isinstance(message, FlowRunInfo):
-                self._storage.persist_flow_run(message)
-                return message
-            elif isinstance(message, NodeRunInfo):
-                self._storage.persist_node_run(message)
-                return message
-        except queue.Empty:
-            pass
-        return None
+    # region monitor thread target function
 
     def _monitor_workers_and_process_tasks_in_thread(
         self,
         task_queue: Queue,
-        result_list: List[LineResult],
+        result_dict: Dict[int, LineResult],
         index: int,
         input_queue: Queue,
         output_queue: Queue,
-        batch_start_time: datetime,
     ):
-        index, process_id, process_name = self._get_process_info(index)
-
-        # Entering the while loop requires two conditions:
-        # 1. The task queue is not empty, meaning there are lines yet to be executed.
-        # 2. The batch run has not reached the batch timeout limit.
-        while not self._batch_timeout_expired(batch_start_time):
-            self._processes_manager.ensure_healthy()
-            try:
-                # Get task from task_queue
-                inputs, line_number, run_id = task_queue.get(timeout=1)
-            except queue.Empty:
-                break
+        # Get the process info of the thread monitoring from the manager.
+        index, process_id, process_name = self._processes_manager.get_process_info(index)
+        self._processes_manager.ensure_healthy()
```

Comment:
Move it[self._processes_manager.ensure_healthy()] to the while loop below.

Reply:
Updated.